added isostress averaging to CompositeViscoPlastic

bobmyhill opened the pull request:
This PR implements isostress (Reuss) component averaging to the CompositeViscoPlastic rheology. 

Aside from giving users the option of choosing from two averaging schemes, isostress averaging is the easiest way to implement a logically consistent viscoelastic model. This is because multiple elastic elements can be combined into a single element, and the "pseudoviscous" part of the elastic element (see Moresi et al., 2003) can be combined with the other viscous elements. Therefore, #5984 should build on this PR.

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).
* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have [created a testcase](https://aspect-documentation.readthedocs.io/en/latest/user/extending/testing/writing-tests.html) for the new feature/benchmark in the [tests/](../blob/main/tests/) directory.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

bobmyhill commented:
/rebuild

gassmoeller commented on source/material_model/rheology/composite_visco_plastic.cc:
```diff
@@ -82,38 +82,302 @@ namespace aspect
+            if (use_diffusion_creep)
+              {
+                log_edot_and_deriv[i][0] = diffusion_creep->compute_log_strain_rate_and_derivative(log_viscoplastic_stress, pressure, temperature, grain_size, diffusion_creep_parameters[i]);
+                log_edot_and_deriv[i][0].first += log_volume_fraction;
```
why do you add the log_volume_fraction to the log strain rate? is it because the addition in log-space is the same as the multiplication in without the log? please add a comment to explain.

gassmoeller commented on source/material_model/rheology/composite_visco_plastic.cc:
```diff
@@ -82,38 +82,302 @@ namespace aspect
+        // between these components.
+
+        // The following while loop contains a Newton iteration to obtain the
+        // viscoplastic stress that is consistent with the total strain rate.
```
please add a comment saying that this also results in a correct partitioning between the flow mechanisms (the introductory sentences of the comment refer to the partitioning, but the last sentence doesnt explain that the Newton iteration will result in a correct partitioning).

gassmoeller commented on source/material_model/rheology/composite_visco_plastic.cc:
```diff
@@ -324,10 +580,25 @@ namespace aspect
       CompositeViscoPlastic<dim>::declare_parameters (ParameterHandler &prm)
       {
+        prm.declare_entry ("Viscosity averaging scheme", "isostress",
+                           Patterns::Selection("isostress|isostrain|unspecified"),
```
no need to have an unspecified option if it is not working and not the default

gassmoeller reviewed (commented):
Very nice :+1:. A few comments on the structure of the interface (public vs private) and naming (isostress vs composition), but otherwise just smaller comments.

bobmyhill commented on include/aspect/material_model/rheology/composite_visco_plastic.h:
```diff
@@ -107,11 +135,19 @@ namespace aspect
-          calculate_log_strain_rate_and_derivative(const std::array<std::pair<double, double>, 4> &logarithmic_strain_rates_and_stress_derivatives,
-                                                   const double viscoplastic_stress,
-                                                   std::vector<double> &partial_strain_rates) const;
+          calculate_composition_log_strain_rate_and_derivative(const std::array<std::pair<double, double>, 4> &logarithmic_strain_rates_and_stress_derivatives,
```
I'd also like this symmetry, but I don't see a way to do it neatly.

The problem: while the isostress viscosity solves for a single stress, the isostrain rheology solves for one stress per compositional field. It felt natural to create inner functions that only do one solve, leading to the two functions for "composition" (used several times by the isostrain branch) and the two functions for "isostress" (only used once by the isostress branch).  

I'm happy to group the loop over compositions into a `compute_isostrain_viscosity` if you think that would be neater, but that would still leave a `calculate_composition_log_strain_rate_and_derivative`.

What would you prefer?

gassmoeller replied:
Thanks for the explanation, I understand now. Yes outsourcing the loop over compositions into a `compute_isostrain_viscosity` and then mentioning in the `calculate_composition...` functions that they are only computing things for one composition would be most intuitive I think. What confused me most about the current situation is that there was `compute_isostress` but no `compute_isostrain`, but instead `compute_composition_...`.

bobmyhill replied:
Great, I've now made a new compute_isostrain_viscosity function. I agree that this makes more logical sense.

bobmyhill commented:
@gassmoeller , brilliant, thanks for the comments. I've updated the code as suggested and modified the tests to use the new structure (the functions that they tested are now hidden).

I haven't yet tested the use of phases, so I've removed them from the test cc files. I'll add more tests for phases in a later PR.

This should be reasy for another review now.

gassmoeller reviewed (approved):
Yes, I like the new structure. Good to go :+1:
